### DISCUSSION TITLE
[Question]: Transactions collection explained

### DISCUSSION THREAD

**Discussion by igodsewer** [danny-avila/LibreChat#1640] [Q&A] [answered]
### What is your question?

Hi everybody!
I was looking around and could not find a good explanation of structure of objects stored in **transactions** collection in Mongo. 

There's an example object from my collection and some lines are commented with questions. Can somebody please help and tell me what these mean and how do they correspond with stuff I see on OpenAI dashboard?
PS - All actual data is cut from the paste.
```
{
    "_id": {
        "$oid": "some_id"
    },
    "user": {
        "$oid": "some_id"
    },
    "conversationId": "some_id",
    "tokenType": "completion",
    "model": "model",
    "context": "message",
    "rawAmount": -000, // what's this?
    "tokenValue": -00000, // what's this?
    "rate": 00, // what's this?
    "createdAt": {
        "$date": "some_date"
    },
    "updatedAt": {
        "$date": "some_date"
    },
    "__v": 0
}
```

I'm asking this to try to allocate per-user costs, as I'm using this in a kind of enterprise environment.

Many thanks!

### More Details

No more details :-(

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment, marked as answer]
> "rawAmount": -000, // what's this?

Raw amount of tokens as counted per the tokenizer algorithm.

>    "tokenValue": -00000, // what's this?

Token credits value. 1000 credits = $0.001 (1 mill USD)

> "rate": 00, // what's this?

The rate at which tokens are charged as credits. 

For example, gpt-3.5-turbo-1106 has a rate of 1 for user prompt (input) and 2 for completion (output)

| Model                 | Input                | Output               |
|-----------------------|----------------------|----------------------|
| gpt-3.5-turbo-1106    | $0.0010 / 1K tokens  | $0.0020 / 1K tokens  |


Given the provided example:

    "rawAmount": -137
    "tokenValue": -205.5
    "rate": 1.5

$` \text{Token Value} = (\text{Raw Amount of Tokens}) \times (\text{Rate}) `$

$` 137 \times 1.5 = 205.5 `$

And to get the real amount of USD spend based on **Token Value**:

$` \frac{\text{Token Value}}{1,000,000} = \left(\frac{\text{Raw Amount of Tokens} \times \text{Rate}}{1,000,000}\right) `$

$` \frac{205.5}{1,000,000} = \$0.0002055 \text{ USD} `$

The relevant file for editing rates is found in `api/models/tx.js`

There will be more customization for this soon from the `librechat.yaml` file.

**Reply by igodsewer** [reply]
Thanks for the clarification! I really appreciate this <3

**Reply by danny-avila** [reply]
you're welcome! hope that all made sense

**Reply by falquaddoomi** [reply]
The explanation makes sense, thanks; out of curiosity, why are the values aside from `rate` negative?
